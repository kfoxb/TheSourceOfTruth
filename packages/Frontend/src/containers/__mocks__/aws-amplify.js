export default class Auth {
  username = 'user'

  password = 'password'

  signIn(username, password) {
    return new Promise((resolve, reject) => {
      if (username === this.username && password === this.password) {
        return resolve();
      }
      return reject();
    });
  }

  currentAuthenticatedUser() {
    return Promise.resolve({ username: this.username });
  }
}
