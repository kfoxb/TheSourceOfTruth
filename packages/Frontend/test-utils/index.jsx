/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    const wrapper = shallow(<Component {...props} />);
    expect(wrapper).toMatchSnapshot();
  };
