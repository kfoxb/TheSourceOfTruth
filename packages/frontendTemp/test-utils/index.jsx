/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const makeShallowCreateWrapper = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    return shallow(<Component {...props} />);
  };

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) => {
  const createWrapper = makeShallowCreateWrapper(Component, defaultProps);
  return (extraProps) => {
    expect(createWrapper(extraProps)).toMatchSnapshot();
  };
};
