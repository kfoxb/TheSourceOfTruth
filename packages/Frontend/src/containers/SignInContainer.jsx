import React, { Component } from 'react';
import { Auth } from 'aws-amplify';
import { connect } from 'react-redux';
import { Redirect, withRouter } from 'react-router';
import PropTypes from 'prop-types';
import SignIn from '../components/SignIn';
import login from '../actions/login';

class SignInContainer extends Component {
  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    login: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props);
    this.state = {
      error: '',
      loading: false,
      password: '',
      username: '',
    };
  }

  updateFormByKey = key => ({ target }) => this.setState({ [key]: target.value });

  signIn = () => {
    const { username, password } = this.state;
    this.setState({ error: '', loading: true }, () => {
      Auth.signIn(username, password)
        .then(() => Auth.currentAuthenticatedUser())
        .then(user => this.props.login(user))
        .catch(error => this.setState({ error }))
        .finally(() => this.setState({ loading: false }));
    });
  }

  submitIfEnter = ({ key }) => {
    if (key === 'Enter') {
      this.signIn();
    }
  }

  render() {
    if (this.props.isAuthenticated) {
      return <Redirect to="/" />;
    }

    const { error, loading } = this.state;
    let errMessage;
    if (error) { errMessage = typeof error === 'string' ? error : error.message; }
    return (<SignIn
      error={errMessage || false}
      loading={loading}
      signIn={this.signIn}
      submitIfEnter={this.submitIfEnter}
      updateFormByKey={this.updateFormByKey}
    />);
  }
}

const mapStateToProps = state => ({
  isAuthenticated: state.user.isAuthenticated,
});

const mapDispatchToProps = dispatch => ({
  login: user => dispatch(login(user)),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(SignInContainer));
