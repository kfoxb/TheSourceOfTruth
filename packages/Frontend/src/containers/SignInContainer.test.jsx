/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { Auth } from 'aws-amplify';
import { createAssertWithPropsToMatchSnapshot } from '../../test-utils';
import SignInContainer from './SignInContainer';

jest.mock('aws-amplify', () => ({
  Auth: {
    currentAuthenticatedUser: jest.fn(() => Promise.resolve({ username: 'user' })),
    signIn: jest.fn((username, password) => new Promise((resolve, reject) => {
      if (username === 'user' && password === 'password') {
        resolve();
      }
      reject();
    })),
  },
}));

describe('SignInContainer', () => {
  const defaultProps = {
    isAuthenticated: false,
    login: () => {},
  };
  const assertWithPropsToMatchSnapshot =
    createAssertWithPropsToMatchSnapshot(SignInContainer.WrappedComponent, defaultProps);
  it('should render', () => {
    assertWithPropsToMatchSnapshot();
  });

  it('should render a Redirect when authenticated', () => {
    assertWithPropsToMatchSnapshot({ isAuthenticated: true });
  });

  describe('submitIfEnter', () => {
    it('should call SignIn if the key is Enter', () => {
      const inst = shallow(<SignInContainer.WrappedComponent {...defaultProps} />).instance();
      inst.signIn = jest.fn();
      inst.submitIfEnter({ key: 'Enter' });
      expect(inst.signIn).toHaveBeenCalledTimes(1);
    });

    it('should not call SignIn if the key is not Enter', () => {
      const inst = shallow(<SignInContainer.WrappedComponent {...defaultProps} />).instance();
      inst.signIn = jest.fn();
      inst.submitIfEnter({ key: 'J' });
      expect(inst.signIn).toHaveBeenCalledTimes(0);
    });
  });

  describe('updateFormByKey', () => {
    it('should return a function', () => {
      const inst = shallow(<SignInContainer.WrappedComponent {...defaultProps} />).instance();
      const res = inst.updateFormByKey('key');
      expect(typeof res).toBe('function');
    });

    it('should return a function that calls setState with the key and target', () => {
      const key = 'key';
      const value = 'val';
      const inst = shallow(<SignInContainer.WrappedComponent {...defaultProps} />).instance();
      inst.setState = jest.fn();
      const res = inst.updateFormByKey(key);
      res({ target: { value } });
      expect(inst.setState).toHaveBeenCalledWith({ key: value });
    });
  });

  describe('signIn', () => {
    it('should call setState with { error: \'\', loading: true } and a callback', () => {
      const inst = shallow(<SignInContainer.WrappedComponent {...defaultProps} />).instance();
      inst.setState = jest.fn();
      inst.signIn();
      expect(inst.setState.mock.calls[0][0]).toEqual({ error: '', loading: true });
      expect(typeof inst.setState.mock.calls[0][1]).toBe('function');
    });

    describe('callback', () => {
      it('should call Auth.signIn, Auth.currentAuthenticatedUser, props.login, then set loading false', () => {
        const login = jest.fn();
        const wrapper =
          shallow(<SignInContainer.WrappedComponent {...defaultProps} login={login} />);
        const inst = wrapper.instance();
        wrapper.setState({ username: 'user', password: 'password' });
        inst.setState = jest.fn();
        inst.signIn();
        const callback = inst.setState.mock.calls[0][1];
        callback();
        expect(Auth.signIn).toHaveBeenCalledWith('user', 'password');
      });
    });
  });
});
