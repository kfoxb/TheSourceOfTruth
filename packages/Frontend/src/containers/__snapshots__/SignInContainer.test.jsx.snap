// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInContainer should render 1`] = `
<SignIn
  error={false}
  loading={false}
  signIn={[Function]}
  submitIfEnter={[Function]}
  updateFormByKey={[Function]}
/>
`;

exports[`SignInContainer should render a Redirect when authenticated 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;
